File: cypha_som/gria_controller.py
"""GRIA alpha live topology controller (Upgrade U3)."""

from __future__ import annotations

from collections import deque
from typing import Any, Callable, Deque, Optional

import numpy as np


def _entropy_matrix(X: np.ndarray, n_bins: int = 16) -> float:
    X = np.asarray(X, dtype=np.float64)
    if X.size == 0:
        return 0.0
    flat = X.ravel()
    hist, _ = np.histogram(flat, bins=n_bins)
    hist = hist.astype(np.float64) + 1e-12
    hist /= hist.sum()
    return float(-np.sum(hist * np.log(hist)))


class GRIAController:
    def __init__(
        self,
        window: int = 200,
        low: float = 0.35,
        high: float = 0.65,
        delta_ssm: float = 0.01,
        control_interval: int = 50,
    ) -> None:
        self.window = int(window)
        self.low = float(low)
        self.high = float(high)
        self.delta_ssm = float(delta_ssm)
        self.control_interval = int(control_interval)
        self._inp_buf: Deque[float] = deque(maxlen=window)
        self._act_buf: Deque[float] = deque(maxlen=window)
        self._step = 0

    def push(self, x: np.ndarray, activations: np.ndarray) -> None:
        x = np.asarray(x, dtype=np.float64).ravel()
        a = np.asarray(activations, dtype=np.float64).ravel()
        self._inp_buf.append(float(np.std(x) + 1e-9))
        self._act_buf.append(float(np.std(a) + 1e-9))

    def alpha(self) -> float:
        if len(self._inp_buf) < min(32, self.window):
            return 0.5
        H_x = _entropy_matrix(np.array(self._inp_buf, dtype=np.float64))
        H_f = _entropy_matrix(np.array(self._act_buf, dtype=np.float64))
        return float(np.clip(1.0 - H_f / (H_x + 1e-9), 0.0, 1.0))

    def act(
        self,
        node_id: int,
        gng: Any,
        ssm_adjust: Optional[Callable[[float], None]] = None,
    ) -> str:
        self._step += 1
        if self._step % self.control_interval != 0:
            return "skip"
        a = self.alpha()
        action = "hold"
        if a < self.low:
            gng.force_insert(node_id)
            action = "split"
            if ssm_adjust is not None:
                ssm_adjust(+self.delta_ssm)
        elif a > self.high:
            gng.merge_with_nearest(node_id)
            action = "merge"
            if ssm_adjust is not None:
                ssm_adjust(-self.delta_ssm)
        return action

File: native/src/som/gria_controller.cpp
#include "cypha/som/gria_controller.hpp"

#include "cypha/som/gng_expert.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace cypha::som {

namespace {

constexpr double kEps = 1e-9;
constexpr double kHistEps = 1e-12;

}  // namespace

GRIAController::GRIAController(GRIAControllerConfig cfg)
    : window_(std::max(1, cfg.window)),
      low_(cfg.low),
      high_(cfg.high),
      delta_ssm_(cfg.delta_ssm),
      control_interval_(std::max(1, cfg.control_interval)) {}

double GRIAController::std_dev(const std::vector<double>& v) {
  if (v.empty()) {
    return 0.0;
  }
  const double mean =
      std::accumulate(v.begin(), v.end(), 0.0) / static_cast<double>(v.size());
  double var = 0.0;
  for (double x : v) {
    const double d = x - mean;
    var += d * d;
  }
  var /= static_cast<double>(v.size());
  return std::sqrt(var);
}

double GRIAController::entropy_hist(const std::deque<double>& buf, int n_bins) {
  if (buf.empty()) {
    return 0.0;
  }
  const int bins = std::max(1, n_bins);
  double lo = buf.front();
  double hi = buf.front();
  for (double v : buf) {
    lo = std::min(lo, v);
    hi = std::max(hi, v);
  }
  if (hi <= lo) {
    hi = lo + 1.0;
  }
  std::vector<double> hist(static_cast<std::size_t>(bins), 0.0);
  const double width = (hi - lo) / static_cast<double>(bins);
  for (double v : buf) {
    int idx = static_cast<int>((v - lo) / width);
    if (idx >= bins) {
      idx = bins - 1;
    }
    if (idx < 0) {
      idx = 0;
    }
    hist[static_cast<std::size_t>(idx)] += 1.0;
  }
  double ent = 0.0;
  for (double c : hist) {
    const double p = (c + kHistEps) /
                     (static_cast<double>(buf.size()) + kHistEps * static_cast<double>(bins));
    ent -= p * std::log(p);
  }
  return ent;
}

void GRIAController::push(const std::vector<double>& x,
                          const std::vector<double>& activations) {
  const double sx = std_dev(x) + kEps;
  const double sa = std_dev(activations) + kEps;
  inp_buf_.push_back(sx);
  act_buf_.push_back(sa);
  while (static_cast<int>(inp_buf_.size()) > window_) {
    inp_buf_.pop_front();
  }
  while (static_cast<int>(act_buf_.size()) > window_) {
    act_buf_.pop_front();
  }
}

double GRIAController::alpha() const {
  const int warm = std::min(32, window_);
  if (static_cast<int>(inp_buf_.size()) < warm) {
    return 0.5;
  }
  const double h_x = entropy_hist(inp_buf_);
  const double h_f = entropy_hist(act_buf_);
  const double a = 1.0 - h_f / (h_x + kEps);
  return std::clamp(a, 0.0, 1.0);
}

std::string GRIAController::act(int node_id, GNGExpertManager& gng,
                                std::function<void(double)> ssm_adjust) {
  ++step_;
  if (step_ % control_interval_ != 0) {
    return "skip";
  }
  const double a = alpha();
  std::string action = "hold";
  if (a < low_) {
    gng.force_insert(node_id);
    action = "split";
    if (ssm_adjust) {
      ssm_adjust(+delta_ssm_);
    }
  } else if (a > high_) {
    gng.merge_with_nearest(node_id);
    action = "merge";
    if (ssm_adjust) {
      ssm_adjust(-delta_ssm_);
    }
  }
  return action;
}

}  // namespace cypha::som
